addons/l10n_in_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,573 @@
+                    error_message.append(_(
+                        "Invalid HSN Code (%s) in product %s", hsn_code, line.product_id.name
+                    ))
+            else:
+                error_message.append(_("product is required to get HSN code"))
+        return error_message
+
+    def _l10n_in_edi_get_iap_buy_credits_message(self, company):
+        base_url = "https://iap-sandbox.odoo.com/iap/1/credit" if not company.sudo().l10n_in_edi_production_env else ""
+        url = self.env["iap.account"].get_credits_url(service_name="l10n_in_edi", base_url=base_url)
+        return markupsafe.Markup("""<p><b>%s</b></p><p>%s</p>""" % (
+            _("You have insufficient credits to send this document!"),
+            _("Please proceed to buy more credits <a href='%s'>here.</a>", url),
+        ))
+
+    def _post_invoice_edi(self, invoices):
+        if self.code != "in_einvoice_1_03":
+            return super()._post_invoice_edi(invoices)
+        response = {}
+        res = {}
+        generate_json = self._l10n_in_edi_generate_invoice_json(invoices)
+        response = self._l10n_in_edi_generate(invoices.company_id, generate_json)
+        if response.get("error"):
+            error = response["error"]
+            error_codes = [e.get("code") for e in error]
+            if "1005" in error_codes:
+                # Invalid token eror then create new token and send generate request again.
+                # This happen when authenticate called from another odoo instance with same credentials (like. Demo/Test)
+                authenticate_response = self._l10n_in_edi_authenticate(invoices.company_id)
+                if not authenticate_response.get("error"):
+                    error = []
+                    response = self._l10n_in_edi_generate(invoices.company_id, generate_json)
+                    if response.get("error"):
+                        error = response["error"]
+                        error_codes = [e.get("code") for e in error]
+            if "2150" in error_codes:
+                # Get IRN by details in case of IRN is already generated
+                # this happens when timeout from the Government portal but IRN is generated
+                response = self._l10n_in_edi_get_irn_by_details(invoices.company_id, {
+                    "doc_type": invoices.move_type == "out_refund" and "CRN" or "INV",
+                    "doc_num": invoices.name,
+                    "doc_date": invoices.invoice_date and invoices.invoice_date.strftime("%d/%m/%Y") or False,
+                })
+                if not response.get("error"):
+                    error = []
+                    odoobot = self.env.ref("base.partner_root")
+                    invoices.message_post(author_id=odoobot.id, body=_(
+                        "Somehow this invoice had been submited to government before." \
+                        "<br/>Normally, this should not happen too often" \
```

Comment:
this doesn't sound reasuring :sweat_smile:

Reply:
@mart-e this is not reassuring but this happens when the government server is down and gives a timeout error but data is processed at the government server so we need to call another API call to get data.